mount/mount_linux.go

```diff
@@ -256,6 +257,100 @@ func (mounter *Mounter) GetMountRefs(pathname string) ([]string, error) {
+		case isExitError && ee.ExitStatus() > fsckErrorsUncorrected:
+			klog.Infof("`fsck` error %s", string(out))
+		}
+	}
+	return nil
+}
+
+// checkAndRepairXfsFilesystem checks and repairs xfs filesystem using command xfs_repair.
+func (mounter *SafeFormatAndMount) checkAndRepairXfsFilesystem(source string) error {
+	klog.V(4).Infof("Checking for issues with xfs_repair on disk: %s", source)
+
+	args := []string{source}
+	checkArgs := []string{"-n", source}
+
+	// check-only using "xfs_repair -n", if the exit status is not 0, perform a "xfs_repair"
+	out, err := mounter.Exec.Command("xfs_repair", checkArgs...).CombinedOutput()
+	if err != nil {
+		if err == utilexec.ErrExecutableNotFound {
+			klog.Warningf("'xfs_repair' not found on system; continuing mount without running 'xfs_repair'.")
+			return nil
+		} else {
+			klog.V(4).Infof("Filesystem corruption was detected for %s, running xfs_repair to repair", source)
+			out, err = mounter.Exec.Command("xfs_repair", args...).CombinedOutput()
+			err = analyzeXfsRepairOutput(out)
+			if err != nil {
+				return fmt.Errorf("'xfs_repair' found errors on device %s but could not correct them: %v", source, err)
+			} else {
+				klog.Infof("Device %s has errors which were corrected by xfs_repair.", source)
+				return nil
+			}
+		}
+	}
+	return nil
+}
+
+// analyzeXfsRepairOutput analyzes the output of xfs_repair command to see if
+// there is any errors, becasue the exit status is always 0.
+// a good xfs_repair looks like the following:
+//
+// Phase 1 - find and verify superblock...
+// Phase 2 - using internal log
+//         - zero log...
+//         - scan filesystem freespace and inode maps...
+//         - found root inode chunk
+// Phase 3 - for each AG...
+//         - scan and clear agi unlinked lists...
+//         - process known inodes and perform inode discovery...
+//         - agno = 0
+//         - agno = 1
+//         - agno = 2
+//         - agno = 3
+//         - process newly discovered inodes...
+// Phase 4 - check for duplicate blocks...
+//         - setting up duplicate extent list...
+//         - check for inodes claiming duplicate blocks...
+//         - agno = 0
+//         - agno = 3
+//         - agno = 2
+//         - agno = 1
+// Phase 5 - rebuild AG headers and trees...
+//         - reset superblock...
+// Phase 6 - check inode connectivity...
+//         - resetting contents of realtime bitmap and summary inodes
+//         - traversing filesystem ...
+//         - traversal finished ...
+//         - moving disconnected inodes to lost+found ...
+// Phase 7 - verify and correct link counts...
+// done
+//
+// If xfs_repair is not able to repair the filesystem at a certain phase, it
+// will print some error messages and exit.
+func analyzeXfsRepairOutput(output []byte) error {
+	trimed := bytes.TrimSpace(output)
+	if bytes.HasSuffix(trimed, []byte("done")) {
+		return nil
+	} else {
+		return fmt.Errorf(string(output))
+	}
```

Comment:
I wonder if this could result in breaking existing stuff. I am unsure about searching for "done" string at the end of the message and using combined output from `xfs_repair`. We have been burnt by such checks in past and command output itself may change across versions.

I think may be we should allow different CSI driver authors to opt-in for doing filesystem repair before mounting. We did not call `xfs_repair` before and now we do and we are performing checks based on output of command(which can spew random stuff and provides no guarantees about API) - could this cause breaking behaviour?

Reply:
@gnufied , thank you for pointing it out. I checked several versions of `xfs_repair`, the output format is stable, so I think it is safe to search for "done" here. And I think it won't break existing stuff. Regarding "We have been burnt by such checks in past", do you remember the code and how did you fix it? can you show me so I can learn something.

I don't think we should let the CSI plugins to do the fs check. Because it is a common task, each type of filesystem has a specific check/repair method. Another reason is that `xfs_repair` must be run in a formatted disk, if we let the CSI plugins to run it, they must run `blkid` first, which is redundant.

Yes, for a xfs filesystem, We did not call `xfs_repair` before and now we do. But it won't break anything:
1. If `xfs_repair` is called before, the fs is repaired before enter `mount`,so `xfs_repair -n` will return ok and nothing will be repaireed again.
2. If `xfs_repair` is not called before, we will try to repair it before mount, it will make the `mount` method more robust.

What do you think?

Reply:
It is okay to rely on following outputs of a program called via exec:

1. exit code
2. if program supports some kind of json or other machine readable output.

It is not okay to rely on human readable text printed on terminal for consumption by another program. Because in that case - called program(`xfs_repair` in this case) makes no guarantees about an stable format.

Reply:
@gnufied , after checking the source code `git://git.kernel.org/pub/scm/fs/xfs/xfsprogs-dev.git`, I find that xfs_repair do return a non-zero status if it is failed to repair the filesystem.
The code is in `repair/xfs_repair.c`.
In fact, the problem is the man page, it is wrong until v4.10.0, you can see the changelog of v4.10.0: `https://abi-laboratory.pro/?view=changelog&l=xfsprogs&v=5.0.0` which fixed the description.

So we don't need to analysis the human readable output here, just check the return status

Reply:
attached the diff here to make it more clear.
```
$ git diff v4.9.0 v4.10.0 man/man8/xfs_repair.8
diff --git a/man/man8/xfs_repair.8 b/man/man8/xfs_repair.8
index 1b4d9e3..85e4dc9 100644
--- a/man/man8/xfs_repair.8
+++ b/man/man8/xfs_repair.8
@@ -65,7 +65,9 @@ Forces
 .B xfs_repair
 to zero the log even if it is dirty (contains metadata changes).
 When using this option the filesystem will likely appear to be corrupt,
-and can cause the loss of user files and/or data.
+and can cause the loss of user files and/or data.  See the
+.B "DIRTY LOGS"
+section for more information.
 .TP
 .BI \-l " logdev"
 Specifies the device special file where the filesystem's external
@@ -505,11 +507,38 @@ This message refers to a large directory.
 If the directory were small, the message would read "junking entry ...".
 .SH EXIT STATUS
 .B xfs_repair \-n
-(no modify node)
+(no modify mode)
 will return a status of 1 if filesystem corruption was detected and
 0 if no filesystem corruption was detected.
 .B xfs_repair
-run without the \-n option will always return a status code of 0.
+run without the \-n option will always return a status code of 0 if
+it completes without problems.  If a runtime error is encountered
+during operation, it will return a status of 1.  In this case,
+.B xfs_repair
+should be restarted.  If
+.B xfs_repair is unable
+to proceed due to a dirty log, it will return a status of 2.  See below.
+.SH DIRTY LOGS
+Due to the design of the XFS log, a dirty log can only be replayed
+by the kernel, on a machine having the same CPU architecture as the
+machine which was writing to the log.
+.B xfs_repair
+cannot replay a dirty log and will exit with a status code of 2
+when it detects a dirty log.
+.PP
+In this situation, the log can be replayed by mounting and immediately
+unmounting the filesystem on the same class of machine that crashed.
+Please make sure that the machine's hardware is reliable before
+replaying to avoid compounding the problems.
+.PP
+If mounting fails, the log can be erased by running
+.B xfs_repair
+with the -L option.
+All metadata updates in progress at the time of the crash will be lost,
+which may cause significant filesystem damage.
+This should
+.B only
+be used as a last resort.
 .SH BUGS
 The filesystem to be checked and repaired must have been
 unmounted cleanly using normal system administration procedures
```